tests/mpi/hp_step-4.cc

```diff
@@ -0,0 +1,374 @@
+  };
+
+
+
+  template <int dim>
+  class RightHandSide : public Function<dim>
+  {
+  public:
+    RightHandSide()
+      : Function<dim>()
+    {}
+
+    virtual double
+    value(const Point<dim> &p, const unsigned int component = 0) const override;
+  };
+
+
+
+  template <int dim>
+  double
+  RightHandSide<dim>::value(const Point<dim> &p,
+                            const unsigned int /*component*/) const
+  {
+    double return_value = 0.0;
+    for (unsigned int i = 0; i < dim; ++i)
+      return_value += 4.0 * std::pow(p(i), 4.0);
+
+    return return_value;
+  }
+
+
+
+  template <int dim>
+  Step4<dim>::Step4()
+    : communicator(MPI_COMM_WORLD)
+    , triangulation(communicator,
+                    typename Triangulation<dim>::MeshSmoothing(
+                      Triangulation<dim>::smoothing_on_refinement |
+                      Triangulation<dim>::smoothing_on_coarsening))
+    , dof_handler(triangulation)
+    , pcout(std::cout, (Utilities::MPI::this_mpi_process(communicator) == 0))
+  {
+    fe.push_back(FE_Q<dim>(2));
+    fe.push_back(FE_Nothing<dim>());
+
+    q_collection.push_back(QGauss<dim>(fe[0].degree + 1));
+  }
+
+
+
+  template <int dim>
+  void
+  Step4<dim>::make_grid()
+  {
+    GridGenerator::hyper_cube(triangulation, -1, 1);
+
+    if (dim == 3)
+      triangulation.refine_global(3);
+    else
+      triangulation.refine_global(4);
+
+    for (auto &cell : triangulation.active_cell_iterators())
+      if (cell->is_locally_owned())
+        if ((cell->center()[dim - 1] < 0.) && (cell->center()[dim - 1] > -0.2))
+          cell->set_refine_flag();
+
+    triangulation.execute_coarsening_and_refinement();
+  }
+
+
+
+  template <int dim>
+  void
+  Step4<dim>::setup_system()
+  {
+    for (auto &cell : dof_handler.active_cell_iterators())
+      if (cell->is_locally_owned())
+        {
+          Point<dim> center = cell->center();
+
+          if (center[dim - 1] >= 0)
+            cell->set_active_fe_index(1);
+
+          else
+            cell->set_active_fe_index(0);
+        }
+    dof_handler.distribute_dofs(fe);
+
+    locally_owned_dofs = dof_handler.locally_owned_dofs();
+    locally_relevant_dofs.clear();
+    DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+    locally_relevant_dofs.compress();
+
+    constraints.clear();
+
+    constraints.reinit(locally_relevant_dofs);
+
+    DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+
+    DoFTools::make_zero_boundary_constraints(dof_handler, constraints);
+
+#ifdef DEBUG
+    // We did not think about hp constraints on ghost cells yet.
+    // Thus, we are content with verifying their consistency for now.
+    IndexSet locally_active_dofs;
+    DoFTools::extract_locally_active_dofs(dof_handler, locally_active_dofs);
+    AssertThrow(constraints.is_consistent_in_parallel(
+                  dof_handler.locally_owned_dofs_per_processor(),
+                  locally_active_dofs,
+				  communicator,
```

Comment:
The code indentation of this particular line looks odd. Please run the indentation script as described [here](https://www.dealii.org/current/doxygen/deal.II/CodingConventions.html). Otherwise testers of the CI will complain.

Reply:
I've just noticed that the indentation check of `travis-ci` passed. Hmm, that's strange, but I guess we are good as long as the tester does not complain...